src/SignalR/server/Core/src/DefaultHubLifetimeManager.cs

```diff
@@ -238,7 +238,26 @@ public override Task SendGroupsAsync(IReadOnlyList<string> groupNames, string me
             var group = _groups[groupName];
             if (group != null)
             {
-                DefaultHubLifetimeManager<THub>.SendToGroupConnections(methodName, args, group, null, null, ref tasks, ref message, cancellationToken);
+                foreach (var connection in group)
+                {
+                    if (connections == null)
+                    {
+                        connections = new HashSet<string>();
+                    }
+                    connections.Add(connection.Key);
+                }
+            }
+        }
+
+        if (connections != null)
+        {
+            foreach (var connectionId in connections)
+            {
+                if (tasks == null)
+                {
+                    tasks = new List<Task>();
+                }
+                tasks.Add(SendConnectionAsync(connectionId, methodName, args, cancellationToken));
```

Comment:
This call to SendConnectionAsync reserializes the method invocation payload for each connection.

https://github.com/dotnet/aspnetcore/blob/aa2e0dcd4c7aca3d52785fffbfadec91f2a6101c/src/SignalR/server/Core/src/DefaultHubLifetimeManager.cs#L194-L196

Prior to this change, SendToGroupConnections would call CreateSerializedInvocationMessage once, send the same serialized invocation payload to each client. I suspect that will have a far bigger performance impact than the HashSet, although I'd also like to pool that if we move forward with this change.